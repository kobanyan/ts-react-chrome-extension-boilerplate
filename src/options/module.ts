import actionCreatorFactory from 'typescript-fsa';
import { reducerWithInitialState } from 'typescript-fsa-reducers';
const actionCreator = actionCreatorFactory();

export interface State {
  favoriteColor: string;
  likesColor: boolean;
}

export const INITIAL_STATE: State = {
  favoriteColor: 'red',
  likesColor: false,
};

export const updateFavoriteColor = actionCreator<string>('UPDATE_FAVORITE_COLOR');
export const updateLikesColor = actionCreator<boolean>('UPDATE_LIKES_COLOR');

export const reducer = reducerWithInitialState(INITIAL_STATE)
  .case(updateFavoriteColor, (state, favoriteColor) => ({ ...state, favoriteColor }))
  .case(updateLikesColor, (state, likesColor) => ({ ...state, likesColor }))
  .build();
