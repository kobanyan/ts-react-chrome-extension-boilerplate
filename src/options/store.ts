import { Reducer, createStore, combineReducers } from 'redux';
import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';

import { reducer as options, State } from './module';

export type RootState = {
  options: State;
};

const rootReducer: Reducer<RootState | void> = combineReducers({
  options,
});

const persistConfig = {
  key: 'root',
  storage,
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = createStore(persistedReducer);

export const persistor = persistStore(store);
