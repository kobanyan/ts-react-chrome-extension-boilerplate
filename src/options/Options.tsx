import * as React from 'react';
import { Action } from 'redux';
import { connect, Dispatch } from 'react-redux';

import { RootState } from './store';
import { State, updateFavoriteColor, updateLikesColor } from './module';

interface OptionsProps {
  values: State;
  actions: ActionDispatcher;
}

class ActionDispatcher {
  constructor(private dispatch: (action: Action) => void) {}
  updateFavoriteColor(favoriteColor: string) {
    this.dispatch(updateFavoriteColor(favoriteColor));
  }
  updateLikesColor(likesColor: boolean) {
    this.dispatch(updateLikesColor(likesColor));
  }
}

const FavoriteColor: React.SFC<OptionsProps> = props => {
  return (
    <>
      {chrome.i18n.getMessage('optionsFavoriteColor')}
      <select
        id="color"
        value={props.values.favoriteColor}
        onChange={e => props.actions.updateFavoriteColor(e.currentTarget.value)}
      >
        <option value="red">{chrome.i18n.getMessage('optionsColorsRed')}</option>
        <option value="green">{chrome.i18n.getMessage('optionsColorsGreen')}</option>
        <option value="blue">{chrome.i18n.getMessage('optionsColorsBlue')}</option>
        <option value="yellow">{chrome.i18n.getMessage('optionsColorsYellow')}</option>
      </select>
    </>
  );
};

const LikesColor: React.SFC<OptionsProps> = props => {
  return (
    <label>
      <input
        type="checkbox"
        id="like"
        checked={props.values.likesColor}
        onChange={e => props.actions.updateLikesColor(e.currentTarget.checked)}
      />
      {chrome.i18n.getMessage('optionsLikesColor')}
    </label>
  );
};

const Options: React.SFC<OptionsProps> = props => {
  return (
    <section>
      <FavoriteColor {...props} />
      <LikesColor {...props} />
    </section>
  );
};

export const options = connect(
  (state: RootState) => ({ values: state.options }),
  (dispatch: Dispatch<Action>) => ({ actions: new ActionDispatcher(dispatch) })
)(Options);
