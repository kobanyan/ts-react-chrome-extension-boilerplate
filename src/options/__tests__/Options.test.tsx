import * as React from 'react';
import { Provider } from 'react-redux';
import * as renderer from 'react-test-renderer';
import { configure, mount } from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
const chrome = require('sinon-chrome');
const I18nPlugin = require('sinon-chrome/plugins').I18nPlugin;
import { options as Options } from '../Options';
import { INITIAL_STATE } from '../module';
import { store } from '../store';

const connectedOptions = () => {
  return (
    <Provider store={store}>
      <Options />
    </Provider>
  );
};

describe('Options', () => {
  beforeAll(() => {
    configure({ adapter: new Adapter() });
    chrome.registerPlugin(
      new I18nPlugin({
        optionsFavoriteColor: { message: 'mocked: optionsFavoriteColor' },
        optionsColorsRed: { message: 'mocked: optionsColorsRed' },
        optionsColorsGreen: { message: 'mocked: optionsColorsGreen' },
        optionsColorsBlue: { message: 'mocked: optionsColorsBlue' },
        optionsColorsYellow: { message: 'mocked: optionsColorsYellow' },
        optionsLikesColor: { message: 'mocked: optionsLikesColor' },
        optionsSave: { message: 'mocked: optionsSave' },
        optionsStatus: { message: 'mocked: optionsStatus' },
      })
    );
    global.chrome = chrome;
  });
  afterAll(() => {
    chrome.flush();
    delete global.chrome;
  });
  it('should match snapshot', () => {
    const component = renderer.create(connectedOptions());
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should select', () => {
    const options = mount(connectedOptions());
    options.find('select').simulate('change', { target: { value: 'green' } });
    // https://github.com/airbnb/enzyme/issues/218
    // expect(options.state().favoriteColor).toEqual('green');
  });
  it('should check', () => {
    const options = mount(connectedOptions());
    options.find('input').simulate('change', { target: { value: true } });
    // https://github.com/airbnb/enzyme/issues/218
    // expect(options.state().likesColor).toEqual(true);
  });
});
