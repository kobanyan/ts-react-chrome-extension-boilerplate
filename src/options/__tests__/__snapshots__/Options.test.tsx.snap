// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Options should match snapshot 1`] = `
<section>
  mocked: optionsFavoriteColor
  <select
    id="color"
    onChange={[Function]}
    value="red"
  >
    <option
      value="red"
    >
      mocked: optionsColorsRed
    </option>
    <option
      value="green"
    >
      mocked: optionsColorsGreen
    </option>
    <option
      value="blue"
    >
      mocked: optionsColorsBlue
    </option>
    <option
      value="yellow"
    >
      mocked: optionsColorsYellow
    </option>
  </select>
  <label>
    <input
      checked={false}
      id="like"
      onChange={[Function]}
      type="checkbox"
    />
    mocked: optionsLikesColor
  </label>
</section>
`;
